src/gpasswd.1
.\" -*- nroff -*-
.\" Copyright (C) 2003, 2004, 2006 Thorsten Kukuk
.\" Author: Thorsten Kukuk <kukuk@suse.de>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2 as
.\" published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software Foundation,
.\" Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.TH gpasswd 1 "July 2006" "pwdutils"
.SH NAME
gpasswd \- change group password
.SH SYNOPSIS
\fBgpasswd\fR [\fB-r\fR|\fB-l\fR|\fB-u\fR] [\fIgroup\fR]
.SH DESCRIPTION
\fBgpasswd\fR changes passwords for group accounts. Only an
administrator may change the password for any group. The used
hash algorithmus is defined by the value of \fIGROUP_CRYPT\fR
or, if not specified, \fICRYPT\fR from \fI/etc/default/passwd\fR.
If not configured, the traditinal \fIdes\fR algorithmus is used.
.SH OPTIONS
.TP
.B "\-r"
Remove group password.
.TP
.B "\-l"
A system administrator can lock the password of the specified group.
.TP
.B "\-u"
A system administrator can unlock the password of the specified group,
if the group is not passwordless afterwards (it will not unlock a group
that has only  "!" as a password).
.SH FILES
/etc/group \- group account information
/etc/default/passwd \- default values for password hash
.SH SEE ALSO
.BR group (5),
.BR groupadd (8),
.BR groupdel (8),
.BR groupmod (8)
.SH AUTHOR
Thorsten Kukuk <kukuk@suse.de>


NEWS
pwdutils NEWS -- history of user-visible changes.

Copyright (C) 2002-2012 Thorsten Kukuk

Please send bug reports, questions and suggestions to <kukuk@thkukuk.de>.

Version 3.2.19
* Fix passwd -S

Version 3.2.18
* Fix changing of permissions if uid/gid of user will be modified

Version 3.2.17
* Fix useradd.local

Version 3.2.16
* Make sha512 new default password hash

Version 3.2.15
* newgrp: honour dynamically assigned groups

Version 3.2.14
* pwck: check +/- entries, too
* grpck: check */- entries, too
* login.defs: remove LOGIN_RETRIES
 
Version 3.2.13
* Only copy regular files if usermod moves home directory

Version 3.2.12
* Make sure modified group files are written to disk
* Update translations for sv, fr

Version 3.2.11
* Update translations
* Add translation for nl

Version 3.2.10
* Fix encoding of polish translations
* Fix long options of grpck/pwck
 
Version 3.2.9
* Move UMASK from /etc/login.defs to /etc/default/useradd

Version 3.2.8
* Fix usermod manual page
* Add quiet option to pam_rpasswd.so

Version 3.2.7
* Fix possible endless loop in userdel

Version 3.2.6
* Fix rpasswd protocol error
* Update translations

Version 3.2.5
* Add translations for da
* pwconv: Remove temporary files in case of an error

Version 3.2.4
* Make sure changes are really written to disk

Version 3.2.3
* New translations
* Adjust documentation

Version 3.2.2
* Don't lock system accounts by default

Version 3.2.1
* Adjust for new libxcrypt

Version 3.2.0
* Better support for audit plugin
* --preferred-uid/gid option for useradd and groupadd

Version 3.1.4
* Update polish translation.
* Remove FTMP references.
* Better pwconv support for  NIS entries

Version 3.1.3
* Allow compiling without native language support.
* Update swedish translation.

Version 3.1.2
* Check if crontab is installed before calling it.
* Fix seg.fault in pwconv in error case.

Version 3.1.1
* Adjust PAM config files

Version 3.1.0
* gpasswd: Add -l and -u option to lock/unlock group passwords
* passwd: Add support for BioAPI
* Add LDAP support for RFC2307bis

Version 3.0.7
* passwd.1: correct manual page
* passwd: Allow changing passwords with PAM even if service is unknown

Version 3.0.6
* chfn: Fix syntax check for home phone number

Version 3.0.5
* Fix compilation error of PAM module with new Linux-PAM release
* Fix CHARACTER_CLASS in /etc/login.defs

Version 3.0.4
* Fix pam_rpasswd.so

Version 3.0.3
* usermod: Adjust sp_lstchg if -p option is used
* useradd: Preserve existing home directory
* chage: Fix off-by-one errors in output

Version 3.0.2
* Fix loading of logging plugins
* Unify options (use always --service)
* Add -a option to usermod for shadow compatibility

Version 3.0.1
* usermod: Don't remove old home directory if an error occured

Version 3.0
* Add --stdin option for passwd
* Fix useradd -D option
* Fix PAM/syslog interaction
* Fix lot of ignored failing syscalls
* Compile daemons with -fpie/-pie
* Add support for logging plugins
* Add plugin for LAuS support
* Add plugin for audit support
* Add plugin to log with syslog
* Use libnscd
* Add support for GnuTLS
* groupmod: Add options to remove/add single user from a group
* usermod: Add options to remove/add single user from a list of groups
* rpasswd/rpasswdd: Better SLP support

Version 2.6.4
* rpasswd/pam_rpasswd.so: Fix SLP URL parsing

Version 2.6.3
* newgrp: Add -c option to allow sg emulation

Version 2.6.2
* Fix compile problem if SELinux is disabled
* rpasswdd: Re-register SLP service after timeout

Version 2.6.1
* newgrp: Fix NGROUPS_MAX handling

Version 2.6
* Add -f compat flag to groupadd
* Add UTF8/local locale translation
* Make regex for allowed accounts/groups configurable
* Add optional pam_rpasswd module
* Add SLP support to rpasswd/rpasswdd
* Add pwck
* Add grpck
* Add newgrp
* Add /etc/default/passwd support

Version 2.5.1
* Install all manual pages

Version 2.5
* Adjust for libselinux v1.5
* Add pwconv/pwunconv
* Add grpunconv

Version 2.4
* Fix -G handling of usermod
* Fix LDAP support
* Add gpasswd, groupmod and usermod
* Copy always all extended attributes
* Add SELinux support for passwd, chsh and chfn
* Make sure that group, passwd and shadow are always created with
  correct security attributes
* groupadd -g handling fixed
* vipw/vigr added
* Lot of bug fixes
* Add manual pages
* Add groupadd and groupdel
* useradd and userdel can now create/delete LDAP accounts
* Make compile with OpenLDAP 2.0.x again
* Rewrite ldap functions to meet demands
* chpasswd can now change LDAP passwords
* useradd: Lot of new error checks and bug fixes
* chfn/chsh: Allow user to abort with Ctrl-C during data input
* Add test suite

Version 2.3
* Add chpasswd
* Add expiry
* Fix chage options
* rpasswd client now checks server certificate

Version 2.2
* Add "-D binddn" option, to allow modifying LDAP data with a special
  account.
* passwd and chage: Allow modifying of password expiry informations
  stored in a LDAP database.

Version 2.1
* Add --disable-ldap option to configure
* Fix typos in chage.1 manual page
* Rewrite passwd manual page and explain all options

Version 2.0
* Fix -d and -E option of chage
* Use port 774 as new default
* Make compile with glibc 2.3 and gcc 2.95/3.2/3.3

Version 1.99
* Add chage
* Allow modifying of passwd/shadow file not in /etc (only chage yet)

Version 1.98
* Add manual pages for rpasswd and rpasswdd

Version 1.97
* rpasswd now reads config file with server name and port number
* Security fix for rpasswdd: don't allow remote changing of root
  password without authentification
* Add manual page for rpasswd.conf

Version 1.96
* Make compile with gcc 2.x
* root can now change passwords for other user
* timeout if client does not answer fast enough
* Sent LANG as locale to server

Version 1.95
* Make location of ldap config files compile time configurable

Version 1.94
* Add more sanity checks to prevent file corruption

Version 1.93
* Minor bug fixes to rpasswdd

Version 1.92

src/gpasswd.c
/* Copyright (C) 2002, 2003, 2004, 2005, 2006, 2011 Thorsten Kukuk
   Author: Thorsten Kukuk <kukuk@suse.de>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 2 as
   published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

#if defined(HAVE_CONFIG_H)
#include "config.h"
#endif

#include <grp.h>
#include <time.h>
#include <errno.h>
#include <stdio.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <signal.h>
#include <sys/resource.h>
#if defined(HAVE_XCRYPT_H)
#include <xcrypt.h>
#elif defined(HAVE_CRYPT_H)
#include <crypt.h>
#endif
#ifdef HAVE_LIBNSCD_H
#include <libnscd.h>
#endif

#include "i18n.h"
#include "group.h"
#include "logging.h"
#include "logindefs.h"
#include "read-files.h"
#include "error_codes.h"
#include "parse_crypt_arg.h"

#define MAX_PASSWD_TRIES 3

char *
getpass_from_stdin (const char *prompt)
{
  static char password[160]; /* 127 is the longest with current crypt */
  char *ptr;

  password[0] = '\0';
  fprintf (stdout, "%s", prompt);
  fflush (stdout);
  if (fgets (password, sizeof (password), stdin) == NULL)
    {
      fprintf (stderr, "\n");
      fprintf (stderr, _("%s: error reading from stdin!\n"),
	       "gpasswd");
      return NULL;
    }
  fprintf (stdout, "\n");

  /* Remove trailing \n.  */
  ptr = strchr (password, '\n');
  if (ptr)
    *ptr = 0;

  return password;
}

static void
print_usage (FILE *stream, const char *program)
{
  fprintf (stream, _("Usage: %s [-r|-l|-u] group\n"), program);
}

static void
print_help (const char *program)
{
  print_usage (stdout, program);
  fprintf (stdout, _("%s - change group password\n\n"), program);
  fputs (_("  -r             Remove the password for this group\n"), stdout);
  fputs (_("  -l             Locks the password entry for \"group\"\n"),
	 stdout);
  fputs (_("  -u             Try to unlock the password entry for \"group\"\n"),
	 stdout);
  fputs (_("  --service srv  Use nameservice 'srv'\n"), stdout);
  fputs (_("  -D binddn      Use dn \"binddn\" to bind to the LDAP directory\n"),
	 stdout);
  fputs (_("  -P path        Search group file in \"path\"\n"),
	 stdout);
  fputs (_("  --help         Give this help list\n"), stdout);
  fputs (_("  --usage        Give a short usage message\n"), stdout);
  fputs (_("  --version      Print program version\n"), stdout);
  fputs (_("  --stdin        Receive input from stdin instead of /dev/tty\n"), stdout);
  fputs (_("Valid services for -r are: files, nis, nisplus, ldap\n"), stdout);
}

int
main (int argc, char **argv)
{
  const char *program = "gpasswd";
  char *group;
  const char *crypt_str;
  crypt_t use_crypt;
  int remove_password = 0;
  int lock_password = 0;
  int unlock_password = 0;
  char *binddn = NULL;
  int P_flag = 0;
  group_t *gr_data;
  char *use_service = NULL;
  int stdin_flag = 0;

#ifdef ENABLE_NLS
  setlocale(LC_ALL, "");
  bindtextdomain(PACKAGE, LOCALEDIR);
  textdomain(PACKAGE);
#endif

  open_sec_log (program);

  /* Before going any further, raise the ulimit and ignore
     signals.  */
  init_environment ();

  crypt_str = getlogindefs_str ("GROUP_CRYPT", NULL);
  if (crypt_str == NULL)
    crypt_str = getlogindefs_str ("CRYPT", "des");
  use_crypt = parse_crypt_arg (crypt_str);

  /* Parse program arguments */
  while (1)
    {
      int c;
      int option_index = 0;
      static struct option long_options[] =
	{
	  {"remove", required_argument, NULL, 'r'},
	  {"lock", required_argument, NULL, 'l'},
	  {"unlock", required_argument, NULL, 'u'},
	  {"binddn", required_argument, NULL, 'D'},
	  {"path", required_argument, NULL, 'P'},
	  {"stdin", no_argument, NULL, '\251'},
	  {"service", required_argument, NULL, '\252'},
	  {"version", no_argument, NULL, '\255'},
	  {"usage", no_argument, NULL, '\254'},
	  {"help", no_argument, NULL, '\253'},
	  {NULL, 0, NULL, '\0'}
	};

      c = getopt_long (argc, argv, "rluD:P:", long_options,
                       &option_index);
      if (c == EOF)
        break;
      switch (c)
	{
        case 'D':
          binddn = optarg;
          break;
        case 'r':
	  remove_password = 1;
          break;
	case 'l':
	  lock_password = 1;
	  break;
	case 'u':
	  unlock_password = 1;
	  break;
        case '\252':
          if (use_service != NULL)
            {
              print_error (program);
              return E_BAD_ARG;
            }
          if (strcasecmp (optarg, "yp") == 0 ||
              strcasecmp (optarg, "nis") == 0)
            use_service = "nis";
          else if (strcasecmp (optarg, "nis+") == 0 ||
                   strcasecmp (optarg, "nisplus") == 0)
            use_service = "nisplus";
          else if (strcasecmp (optarg, "files") == 0)
            use_service = "files";
#ifdef USE_LDAP
	  else if (strcasecmp (optarg, "ldap") == 0)
	    use_service = "ldap";
#endif
          else
            {
              fprintf (stderr, _("Service `%s' not supported.\n"), optarg);
              print_usage (stderr, program);
              return E_BAD_ARG;
            }
          break;
	case 'P':
	  P_flag = 1;
	  files_etc_dir = strdup (optarg);
	  break;
	case '\253':
          print_help (program);
          return 0;
        case '\255':
          print_version (program, "2006");
          return 0;
        case '\254':
          print_usage (stdout, program);
          return E_USAGE;
	case '\251':
	  stdin_flag = 1;
	  break;
        default:
          print_error (program);
          return E_BAD_ARG;
        }
    }

  argc -= optind;
  argv += optind;

  /* We have more than one groupname. */
  if (argc > 1)
    {
      fprintf (stderr, _("%s: Too many arguments.\n"), program);
      print_error (program);
      return E_USAGE;
    }

  /* We need a group name.  */
  if (argc == 0)
    {
      fprintf (stderr, _("%s: Group argument missing.\n"), program);
      print_error (program);
      return E_USAGE;
    }
  else if (remove_password + lock_password + unlock_password > 1)
    {
      print_error (program);
      return E_USAGE;
    }
  else
    {
      group = argv[0];
      gr_data = find_group_data (group, 0, use_service);
      if (gr_data == NULL || gr_data->service == S_NONE)
        {
          if (use_service)
            fprintf (stderr,
		     _("%s: Group `%s' is not known to service `%s'.\n"),
                     program, group, use_service);
          else
            fprintf (stderr, _("%s: Unknown group `%s'.\n"), program,
		     group);

	  sec_log (program, MSG_UNKNOWN_GROUP, group, getuid ());

          return E_NOPERM;
        }
    }

  /* Only root is allowed to change password for local groups. */
  if ((gr_data->service == S_LOCAL && getuid () != 0) ||
      (gr_data->service == S_LDAP && binddn == NULL) ||
      (gr_data->service != S_LOCAL && gr_data->service != S_LDAP))
    {
      sec_log (program, MSG_PERMISSION_DENIED,
	       gr_data->gr.gr_name, gr_data->gr.gr_gid, getuid ());
      fprintf (stderr, _("%s: Permission denied.\n"), program);
      free_group_t (gr_data);
      return E_NOPERM;
    }

  if (remove_password)
    {
      gr_data->newpassword = strdup ("");
    }
  else if (unlock_password)
    {
      /* If the password is only "!", don't unlock it.  */
      if (gr_data->gr.gr_passwd &&
	  gr_data->gr.gr_passwd[0] == '!' &&
	  strlen (gr_data->gr.gr_passwd) > 1)
        gr_data->newpassword = strdup (&gr_data->gr.gr_passwd[1]);
      else
        {
          fprintf (stderr, _("Cannot unlock the password for `%s'!\n"),
                   gr_data->gr.gr_name);
          free_group_t (gr_data);
          return E_FAILURE;
        }

    }
  else if (lock_password)
    {
      if (gr_data->gr.gr_passwd == NULL)
	gr_data->newpassword = strdup ("!");
      else if (gr_data->gr.gr_passwd[0] != '!')
        {
          gr_data->newpassword =
	    malloc (strlen (gr_data->gr.gr_passwd) + 2);
          if (gr_data->newpassword == NULL)
            return E_FAILURE;
          strcpy (&gr_data->newpassword[1], gr_data->gr.gr_passwd);
          gr_data->newpassword[0] = '!';
        }
      else
        {
          fprintf (stderr, _("Password for `%s' is already locked!\n"),
                   gr_data->gr.gr_name);
          free_group_t (gr_data);
          return E_FAILURE;
        }
    }
  else
    {
      char *p1;
      int try;

      sec_log (program, MSG_CHANGING_GROUP_PASSWORD, group, getuid ());
      fprintf (stdout, _("Changing the password for group %s.\n"), group);

      for (try = 0; try < MAX_PASSWD_TRIES; try++)
	{
	  const char *p2;

	  if(stdin_flag)
	    p1 = getpass_from_stdin (_("New Password: "));
	  else
	    p1 = getpass (_("New Password: "));

	  if (p1 == NULL || *p1 == '\0')
	    {
	      fputs ("\n", stderr);
	    abort_change:
	      sec_log (program, MSG_PASSWORD_CHANGE_ABORTED,
		       group, getuid ());

	      fprintf (stderr, _("Password change aborted.\n"));
	      return E_FAILURE;
	    }

	  /* We need a copy of p1. */
	  p1 = strdup (p1);
	  if (p1 == NULL)
	    {
	      fputs ("running out of memory!\n", stderr);
	      return E_FAILURE;
	    }

	  if(stdin_flag)
	    p2 = getpass_from_stdin(_("Re-enter new password: "));
	  else
	    p2 = getpass (_("Re-enter new password: "));

	  if (p2 == NULL || *p2 == '\0')
	    {
	      fputs ("\n", stderr);
	      free (p1);
	      goto abort_change;
	    }

	  if (strcmp (p1, p2) != 0)
	    {
	      fprintf (stderr, _("Sorry, passwords do not match.\n"));
	      free (p1);
	    }
	  else
	    break;
	}
      if (try == MAX_PASSWD_TRIES)
	{
	  fprintf (stderr, _("%s: Try again later.\n"), program);
	  sec_log (program, MSG_MAX_GROUP_PASSWD_TRIES, group, getuid ());
	  free (p1);
	  goto abort_change;
        }
      else
	{
	  char *salt;
          struct crypt_data output;
          memset (&output, 0, sizeof (output));


          switch (use_crypt)
            {
            case DES:
              /* If we don't support passwords longer 8 characters,
                 truncate them */
              if (strlen (p1) > 8)
                p1[8] = '\0';
              salt =  make_crypt_salt ("", 0);
              if (salt != NULL)
                gr_data->newpassword = strdup (crypt_r (p1, salt, &output));
              else
                {
		  free (p1);
                  fprintf (stderr, _("Cannot create salt for standard crypt"));
		  goto abort_change;
                }
              free (salt);
              break;
            case BLOWFISH:
#if defined(HAVE_XCRYPT_GENSALT_R)
	      /* blowfish has a limit of 72 characters */
	      if (use_crypt == BLOWFISH && strlen (p1) > 72)
		p1[72] = '\0';
              salt = make_crypt_salt ("$2a$", 0 /* XXX crypt_rounds */);
              if (salt != NULL)
                gr_data->newpassword = strdup (crypt_r (p1, salt, &output));
              else
                {
		  free (p1);
                  fprintf (stderr, _("Cannot create salt for blowfish crypt"));
		  goto abort_change;
                }
              free (salt);
              break;
#else
	      fprintf (stderr,
		       _("No support for blowfish compiled in. Using MD5\n"));
#endif
            case MD5:
              /* MD5 has a limit of 127 characters */
              if (strlen (p1) > 127)
                p1[127] = '\0';
              salt = make_crypt_salt ("$1$", 0);
              if (salt != NULL)
                gr_data->newpassword = strdup (crypt_r (p1, salt, &output));
              else
                {
		  free (p1);
                  fprintf (stderr, _("Cannot create salt for MD5 crypt"));
		  goto abort_change;
                }
              free (salt);
              break;
            default:
              abort();
            }
	  free (p1);
        }
    }

  gr_data->todo = DO_MODIFY;
  if (write_group_data (gr_data, 0) != 0)
    {
      fprintf (stderr, _("%s: Error changing password.\n"),
               program);
      sec_log (program, MSG_ERROR_CHANGE_GROUP_PASSWORD, group, getuid ());
      free_group_t (gr_data);
      return E_FAILURE;
    }

  if (strcmp(gr_data->newpassword,"") == 0)
    {
      printf (_("Password removed.\n"));
      sec_log (program, MSG_GROUP_PASSWORD_REMOVED, group,
               gr_data->gr.gr_gid, getuid ());
    }
  else
    {
      printf (_("Password changed.\n"));
      sec_log (program, MSG_GROUP_PASSWORD_CHANGED,
               group, gr_data->gr.gr_gid, getuid ());
    }

#ifdef HAVE_NSCD_FLUSH_CACHE
  nscd_flush_cache ("group");
#endif

  free_group_t (gr_data);

  return E_SUCCESS;
}
